openedx/core/djangoapps/ccxcon/tests/test_api.py

```diff
@@ -0,0 +1,213 @@
+"""
+Unit tests for the API module
+"""
+
+import datetime
+import mock
+import pytz
+import urlparse
+from nose.plugins.attrib import attr
+
+from opaque_keys.edx.keys import CourseKey
+from student.tests.factories import AdminFactory
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.tests.django_utils import (
+    SharedModuleStoreTestCase,
+    TEST_DATA_SPLIT_MODULESTORE)
+from xmodule.modulestore.tests.factories import (
+    CourseFactory,
+    ItemFactory,
+)
+
+from openedx.core.djangoapps.ccxcon import api as ccxconapi
+from .factories import CcxConFactory
+
+
+def flatten(seq):
+    """
+    For [[1, 2], [3, 4]] returns [1, 2, 3, 4].  Does not recurse.
+    """
+    return [x for sub in seq for x in sub]
+
+
+def fetch_token_mock(*args, **kwargs):  # pylint: disable=unused-argument
+    """
+    Mock function used to bypass the oauth fetch token
+    """
+    return
+
+
+@attr('shard_1')
+class APIsTestCase(SharedModuleStoreTestCase):
+    """
+    Unit tests for the API module functions
+    """
+    MODULESTORE = TEST_DATA_SPLIT_MODULESTORE
+
+    @classmethod
+    def setUpClass(cls):
+        super(APIsTestCase, cls).setUpClass()
+        cls.course = course = CourseFactory.create()
+        cls.course_key = cls.course.location.course_key
+
+        # Create a course outline
+        start = datetime.datetime(
+            2010, 5, 12, 2, 42, tzinfo=pytz.UTC
+        )
+        due = datetime.datetime(
+            2010, 7, 7, 0, 0, tzinfo=pytz.UTC
+        )
+
+        cls.chapters = [
+            ItemFactory.create(start=start, parent=course) for _ in xrange(2)
+        ]
+        cls.sequentials = flatten([
+            [
+                ItemFactory.create(parent=chapter) for _ in xrange(2)
+            ] for chapter in cls.chapters
+        ])
+        cls.verticals = flatten([
+            [
+                ItemFactory.create(
+                    start=start, due=due, parent=sequential, graded=True, format='Homework', category=u'vertical'
+                ) for _ in xrange(2)
+            ] for sequential in cls.sequentials
+        ])
```

Comment:
I do not think it is possible to do it; if I try to change the code to something like this:
```
with cls.store.bulk_operations(course.id, emit_signals=False):
            cls.verticals = flatten([
                [
                    ItemFactory.create(
                        start=start, due=due, parent=sequential, graded=True, format='Homework', category=u'vertical'
                    ) for _ in xrange(2)
                ] for sequential in cls.sequentials
            ])
        with cls.store.bulk_operations(course.id, emit_signals=False):
            blocks = flatten([  # pylint: disable=unused-variable
                [
                    ItemFactory.create(parent=vertical) for _ in xrange(2)
                ] for vertical in cls.verticals
            ])
```
I get an error like this:
```
1) ERROR: test suite for <class 'openedx.core.djangoapps.ccxcon.tests.test_api.APIsTestCase'>

   Traceback (most recent call last):
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/nose/suite.py line 209 in run
      self.setUp()
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/nose/suite.py line 292 in setUp
      self.setupContext(ancestor)
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/nose/suite.py line 315 in setupContext
      try_run(context, names)
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/nose/util.py line 471 in try_run
      return func()
    openedx/core/djangoapps/ccxcon/tests/test_api.py line 79 in setUpClass
      ] for sequential in cls.sequentials
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py line 559 in create
      return cls._generate(True, attrs)
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py line 484 in _generate
      obj = cls._prepare(create, **attrs)
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/factory/base.py line 459 in _prepare
      return cls._create(model_class, *args, **kwargs)
    common/lib/xmodule/xmodule/modulestore/tests/factories.py line 415 in _create
      return published_parent.get_children()[-1]
   IndexError: list index out of range
```